src/frequenz/client/microgrid/_exception.py

```diff
@@ -3,6 +3,529 @@
 
 """Exceptions raised by the microgrid API client."""
 
+from __future__ import annotations
+
+from typing import Protocol
+
+import grpclib
+
 
 class ClientError(Exception):
     """There was an error in the microgrid API client."""
+
+    def __init__(
+        self,
+        *,
+        server_url: str,
+        operation: str,
+        description: str,
+    ) -> None:
+        """Create a new instance.
+
+        Args:
+            server_url: The URL of the server that returned the error.
+            operation: The operation that caused the error.
+            description: A human-readable description of the error.
+        """
+        super().__init__(
+            f"Failed calling {operation!r} on {server_url!r}: {description}"
+        )
+
+        self.server_url = server_url
+        """The URL of the server that returned the error."""
+
+        self.operation = operation
+        """The operation that caused the error."""
+
+        self.description = description
+        """The human-readable description of the error."""
+
+    @classmethod
+    def from_grpc_error(
+        cls,
+        *,
+        server_url: str,
+        operation: str,
+        grpc_error: grpclib.GRPCError,
+        retryable: bool = True,
+    ) -> GrpcStatusError:
+        """Create an instance of the appropriate subclass from a gRPC error.
+
+        Args:
+            server_url: The URL of the server that returned the error.
+            operation: The operation that caused the error.
+            grpc_error: The gRPC error to convert.
+
+        Returns:
+            An instance of
+                [GrpcStatusError][frequenz.client.microgrid.GrpcStatusError] if
```

Comment:
Maybe. Do you think it would be useful to catch a `UnrecognizedGrpcError`? Otherwise I'm not sure what would be the advantage of giving it a special class. I can add it anyway, I'm not against it, mostly curious if you already had a use case in mind.

Reply:
oh, didn't realize `GrpcStatusError` is a superclass for all the other errors.  Why is it not just called `GrpcError`, btw?

Reply:
> I can add it anyway, I'm not against it, mostly curious if you already had a use case in mind.

No, it makes sense the way you've done it, thought the `GrpcStatusError` was a separate thing just for the unrecognized case.  But also found the `Status` part of the error's name weird.

Reply:
Yeah, actually because `GrpcStatusError` is a superclass for all other errors it makes sense to have this separately, otherwise there is no way to tell if the status we received is unrecognized, if you catch `GrpcStatusError`, then you'll catch all of the others.

I wondered if it shouldn't be just mapped to `UnknownError`, but I think it shouldn't because so far every know gRPC error status is mapped to a single class, and if we do so `UnknownError` could have a `UNKNOWN` status or any other random `int` that we don't know about as status, which is more confusing, so I guess a dedicated error for an unrecognized status still makes sense.